### ISSUE TITLE
[Question][Hw3] Can I "mutate" bbenv inside extract_bb?

### ISSUE THREAD

**Issue by retroinspect** [prosyslab-classroom/cs492-program-reasoning#66]
Given signature of extract_bb seems to be `bbenv -> Llvm.llbasicblock -> unit`.
But basically we need "updated" bbenv which considered all blocks and instructions to extract queries.
Can I mutate bbenv (which is Map)?

(As far as I know, we consider mutation as a last resort in functional programming)

**Comment by Hhro** [comment]
Yes, you can update the `bbenv`. (if you mean **mutate** as an update.)
However, that's not necessary for this homework.
It's well initialized by the `initialize` function in the `main`.

What you need to do is add rules that describe the reachability condition for basic blocks and add them to the Z3 fixed-point solver. Then, query the solver whether the basic block calling `is_assert_fail` is reachable or not.

Note that `Z3.Fixedpoint.add_rule` add a rule to the global context of Z3, so you don't need to carry it while iterating the basic blocks.

Hope it helps!
